modules/reindex/src/main/java/org/elasticsearch/index/reindex/AbstractAsyncBulkIndexByScrollAction.java

```diff
@@ -40,108 +41,180 @@
-            }
-            IndexRequest index = buildIndexRequest(doc);
-            copyMetadata(index, doc);
-            if (script != null) {
-                if (executableScript == null) {
-                    executableScript = scriptService.executable(script, mainRequest.getScript().getParams());
-                    scriptCtx = new HashMap<>();
-                }
-                if (false == applyScript(index, doc, executableScript, scriptCtx)) {
-                    continue;
+            if (accept(doc)) {
+                RequestWrapper<?> request = apply(buildRequest(doc), doc);
+                if (request != null) {
+                    bulkRequest.add(request.self());
                 }
             }
-            bulkRequest.add(index);
         }
-
         return bulkRequest;
     }
 
     /**
-     * Copies the metadata from a hit to the index request.
+     * Used to accept or ignore a search hit. Ignored search hits will be excluded
+     * from the bulk request. It is also where we fail on invalid search hits, like
+     * when the document has no source but it's required.
      */
-    protected void copyMetadata(IndexRequest index, SearchHit doc) {
-        index.parent(fieldValue(doc, ParentFieldMapper.NAME));
-        copyRouting(index, doc);
-        // Comes back as a Long but needs to be a string
-        Long timestamp = fieldValue(doc, TimestampFieldMapper.NAME);
-        if (timestamp != null) {
-            index.timestamp(timestamp.toString());
-        }
-        Long ttl = fieldValue(doc, TTLFieldMapper.NAME);
-        if (ttl != null) {
-            index.ttl(ttl);
+    protected boolean accept(SearchHit doc) {
+        if (doc.hasSource()) {
+            /*
+             * Either the document didn't store _source or we didn't fetch it for some reason. Since we don't allow the user to
+             * change the "fields" part of the search request it is unlikely that we got here because we didn't fetch _source.
+             * Thus the error message assumes that it wasn't stored.
+             */
+            throw new IllegalArgumentException("[" + doc.index() + "][" + doc.type() + "][" + doc.id() + "] didn't store _source");
         }
+        return true;
+    }
+
+    /**
+     * Build the {@link RequestWrapper} for a single search hit. This shouldn't handle
+     * metadata or scripting. That will be handled by copyMetadata and
+     * apply functions that can be overridden.
+     */
+    protected abstract RequestWrapper<?> buildRequest(SearchHit doc);
+
+    /**
+     * Apply various changes depending of the Reindex action and  the search hit,
+     * from copying search hit metadata (parent, routing, etc) to potentially
+     * transforming the {@link RequestWrapper} completely.
+     */
+    protected RequestWrapper<?> apply(RequestWrapper<?> request, SearchHit doc) {
```

Comment:
I think maybe this doesn't have to be protected any more? Maybe it doesn't even have to be a method any more and you can just call the scriptApplier directly?